File: AirWallex/Airwallex/Sources/UI/AWXCountryListViewController.m
//
//  AWXCountryListViewController.m
//  Airwallex
//
//  Created by Victor Zhu on 2020/1/19.
//  Copyright © 2020 Airwallex. All rights reserved.
//

#import "AWXCountryListViewController.h"
#import "AWXCountry.h"
#import "AWXUtils.h"

@interface AWXCountryListViewController () <UITableViewDelegate, UITableViewDataSource, UISearchBarDelegate>

@property (strong, nonatomic) UISearchBar *searchBar;
@property (strong, nonatomic) UITableView *tableView;
@property (strong, nonatomic) NSArray <AWXCountry *> *countries;
@property (strong, nonatomic) NSArray <AWXCountry *> *matchedCountries;

@end

@implementation AWXCountryListViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"Close", @"Close") style:UIBarButtonItemStylePlain target:self action:@selector(close:)];

    _searchBar = [UISearchBar new];
    _searchBar.delegate = self;
    _searchBar.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:_searchBar];
    
    _tableView = [UITableView new];
    _tableView.dataSource = self;
    _tableView.delegate = self;
    _tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [_tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"CountryCell"];
    [self.view addSubview:_tableView];
    
    NSDictionary *views = @{@"searchBar": _searchBar, @"tableView": _tableView};
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[searchBar]|" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-[searchBar][tableView]-|" options:NSLayoutFormatAlignAllLeft | NSLayoutFormatAlignAllRight metrics:nil views:views]];
    
    self.countries = [AWXCountry allCountries];
    self.matchedCountries = self.countries;
}

#pragma mark - UITableViewDataSource & UITableViewDelegate

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.matchedCountries.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CountryCell" forIndexPath:indexPath];
    AWXCountry *country = self.matchedCountries[indexPath.row];
    cell.textLabel.text = country.countryName;
    if ([self.country.countryName isEqualToString:country.countryName]) {
        UIImageView *imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"tick" inBundle:[NSBundle resourceBundle]]];
        cell.accessoryView = imageView;
    } else {
        cell.accessoryView = nil;
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(countryListViewController:didSelectCountry:)]) {
        self.country = self.matchedCountries[indexPath.row];
        [self.delegate countryListViewController:self didSelectCountry:self.country];
    }
}

#pragma mark - UISearchBarDelegate

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    if (searchText.length > 0) {
        self.matchedCountries = [self.countries filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
            AWXCountry *obj = (AWXCountry *)evaluatedObject;
            return [obj.countryName rangeOfString:searchText options:NSCaseInsensitiveSearch].location != NSNotFound;
        }]];
    } else {
        self.matchedCountries = self.countries;
    }
    [self.tableView reloadData];
}

@end

File: Airwallex/Core/Sources/Internal/AWXCountryListViewController.swift
//
//  AWXCountryListViewController.swift
//  Core
//
//  Created by Tony He (CTR) on 2024/7/26.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation

@objc public protocol AWXCountryListViewControllerDelegate {
    @objc optional func countryListViewController(
        _ controller: AWXCountryListViewController, didSelect country: AWXCountry
    )
}

@objcMembers
@objc
public class AWXCountryListViewController: UIViewController {
    public weak var delegate: AWXCountryListViewControllerDelegate?

    private lazy var searchBar: UISearchBar = {
        let sb = UISearchBar()
        sb.delegate = self
        sb.translatesAutoresizingMaskIntoConstraints = false
        return sb
    }()

    private lazy var table: UITableView = {
        let tv = UITableView()
        tv.dataSource = self
        tv.delegate = self
        tv.translatesAutoresizingMaskIntoConstraints = false
        tv.register(UITableViewCell.classForCoder(), forCellReuseIdentifier: "CountryCell")
        return tv
    }()

    var countries = AWXCountry.allCountries()
    var matchedCountries = AWXCountry.allCountries()
    public var currentCountry: AWXCountry?

    override public func viewDidLoad() {
        super.viewDidLoad()
        setupViews()
    }

    func close() {
        if let nav = navigationController {
            nav.popViewController(animated: true)
        } else {
            dismiss(animated: true)
        }
    }
}

private extension AWXCountryListViewController {
    func setupViews() {
        navigationItem.rightBarButtonItem = UIBarButtonItem(
            title: NSLocalizedString("Close", comment: "Close"), style: .plain, target: self,
            action: #selector(close)
        )

        view.addSubview(searchBar)
        view.addSubview(table)

        let views = ["searchBar": searchBar, "tableView": table]
        view.addConstraints(
            NSLayoutConstraint.constraints(
                withVisualFormat: "H:|[searchBar]|", metrics: nil, views: views
            ))
        view.addConstraints(
            NSLayoutConstraint.constraints(
                withVisualFormat: "V:|-[searchBar][tableView]-|", metrics: nil, views: views
            ))
    }
}

extension AWXCountryListViewController: UITableViewDataSource, UITableViewDelegate {
    public func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
        return matchedCountries.count
    }

    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)
        -> UITableViewCell
    {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CountryCell", for: indexPath)
        let country = matchedCountries[indexPath.row]
        cell.textLabel?.text = country.countryName
        if let currentCountry = currentCountry, currentCountry.countryName == country.countryName {
            let imageView = UIImageView(image: UIImage(named: "tick", in: Bundle.resource()))
            cell.accessoryView = imageView
        } else {
            cell.accessoryView = nil
        }
        return cell
    }

    public func tableView(_: UITableView, didSelectRowAt indexPath: IndexPath) {
        currentCountry = matchedCountries[indexPath.row]
        if let currentCountry = currentCountry {
            delegate?.countryListViewController?(self, didSelect: currentCountry)
        }
    }
}

extension AWXCountryListViewController: UISearchBarDelegate {
    public func searchBar(_: UISearchBar, textDidChange searchText: String) {
        if !searchText.isEmpty {
            matchedCountries = countries.filter {
                $0.countryName.contains(searchText)
            }
        } else {
            matchedCountries = countries
        }
        table.reloadData()
    }
}